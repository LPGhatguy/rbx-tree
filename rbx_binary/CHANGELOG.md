# rbx_binary Changelog

## Unreleased

## 0.6.0-alpha.5 (2021-05-14)
* Added `OptionalCoordinateFrame` support. ([#176][pr-176])
* Added support for writing CFrame basic rotation IDs. ([#171][pr-171])

[pr-171]: https://github.com/rojo-rbx/rbx-dom/pull/171
[pr-176]: https://github.com/rojo-rbx/rbx-dom/pull/176

## 0.6.0-alpha.4 (2021-04-09)
* Unknown property types will no longer fail deserialization ([#168][pr-168])
    * This makes rbx_binary more resilient to new types being added by Roblox.
* `PROP` chunks that end after their prop name are now silently ignored. ([#168][pr-168])
    * This matches Roblox's behavior and fixes our handling of `OptionalCoordinateFrame`, added in the most recent Roblox Studio release.

[pr-168]: https://github.com/rojo-rbx/rbx-dom/pull/168

## 0.6.0-alpha.3 (2021-03-08)
* Added support for encoding `Int32` values as `BrickColor`.
  * This helps improve compatibility when decoding XML models and reencoding them as binary.

## 0.6.0-alpha.2 (2021-03-04)
* Added support for widening integer types when serializing.
    * `Int32` will widen to `Int64`, `Float32` will widen to `Float64`.

## 0.6.0-alpha.1 (2021-02-16)
This release is a major, breaking change that upgrades rbx\_xml's underlying DOM implementation from rbx\_dom\_weak 1.0 to 2.0. This release also realigned rbx\_binary's API to match rbx_xml.

* Breaking: ported crate to rbx\_dom_weak 2.0
* Breaking: updated top-level API to match rbx_xml.
* Added support for all remaining types.

## 0.5.0 (2019-12-18)
0.5.0 is intended to be mostly API-compatible with previous rbx_binary releases because it makes a lot of foundational changes. 0.6.0 will break the rbx_binary API significantly.

* Rewrote crate from ground-up using understanding gained from rbx_xml
* Added support for using reflection information, improving content compatibility
* Improved performance by up to 30% for some files
* Improved instrumentation using the `log` crate
* Improved the crate's error types ([#48](https://github.com/rojo-rbx/rbx-dom/issues/48))
* Fixed panics in many cases, instead returning an error ([#26](https://github.com/rojo-rbx/rbx-dom/issues/26))
* Fixed handling unknown `BinaryString` values ([#49](https://github.com/rojo-rbx/rbx-dom/issues/49))

## 0.4.1 (2019-05-29)
* Fixed bad interaction with rbx_dom_weak 1.6.0 causing instances to go missing.

## 0.4.0 (2019-03-01)
* Updated to `rbx_dom_weak` 1.0

## 0.3.0 (2019-02-14)
* Updated `rbx_tree` dependency to `rbx_dom_weak` 0.3.0

## 0.2.0 (2019-01-25)
* Updated `rbx_tree` dependency to 0.2.0

## 0.1.0
* Initial release
* Supports `String` and `Bool` types
