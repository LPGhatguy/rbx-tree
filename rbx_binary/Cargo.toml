[package]
name = "rbx_binary"
description = "Implementation of Roblox's binary model (rbxm) and place (rbxl) file formats"
version = "0.6.0-alpha.5"
license = "MIT"
documentation = "https://docs.rs/rbx_binary"
homepage = "https://github.com/rojo-rbx/rbx-dom"
repository = "https://github.com/rojo-rbx/rbx-dom.git"
readme = "README.md"
authors = ["Lucien Greathouse <me@lpghatguy.com>"]
edition = "2018"

[features]
unstable_text_format = ["serde"]

[dependencies]
rbx_dom_weak = { version = "2.0.0-alpha.1", path = "../rbx_dom_weak" }
rbx_reflection = { version = "4.0.0-alpha.1", path = "../rbx_reflection" }
rbx_reflection_database = { version = "0.1.0", path = "../rbx_reflection_database" }

log = "0.4.6"
lz4 = "1.23.1"
thiserror = "1.0.16"
serde = { version = "1.0.103", features = ["derive"], optional = true }

[dev-dependencies]
criterion = "0.3"
env_logger = "0.7.1"
heck = "0.3.1"
insta = "0.16.0"
serde = { version = "1.0.103", features = ["derive"] }

[[bench]]
name = "deserializer"
harness = false

[[bench]]
name = "serializer"
harness = false
