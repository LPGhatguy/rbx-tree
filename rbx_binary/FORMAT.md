# Roblox Binary Model Format, Version 0
This document is based on:
- [*ROBLOX File Format* by Gregory Comer](http://www.classy-studios.com/Downloads/RobloxFileSpec.pdf)
- [LibRbxl by Gregory Comer](https://github.com/GregoryComer/LibRbxl)
- [rbxfile by Anaminus](https://github.com/RobloxAPI/rbxfile)
- [Roblox-File-Format by CloneTrooper1019](https://github.com/CloneTrooper1019/Roblox-File-Format)
- Observing `rbxm` and `rbxl` output from Roblox Studio

## Contents
- [Document Conventions](#document-conventions)
- [File Structure](#file-structure)
- [File Header](#file-header)
- [Chunks](#chunks)
	- [`META` Chunk](#meta-chunk)
	- [`SSTR` Chunk](#sstr-chunk)
	- [`INST` Chunk](#inst-chunk)
	- [`PROP` Chunk](#prop-chunk)
	- [`PRNT` Chunk](#prnt-chunk)
	- [`END` Chunk](#end-chunk)
- [Data Types](#data-types)
	- [String](#string)
	- [Bool](#bool)
	- [Int32](#int32)
	- [Float32](#float32)
	- [Float64](#float64)
	- [UDim](#udim)
	- [UDim2](#udim2)
	- [Ray](#ray)
	- [Faces](#faces)
	- [Axis](#axis)
	- [BrickColor](#brickcolor)
	- [Color3](#color3)
	- [Vector2](#vector2)
	- [Vector3](#vector3)
	- [CFrame](#cframe)
	- [Enum](#enum)
	- [Referent](#referent)
	- [Vector3int16](#vector3int16)
	- [NumberSequence](#numbersequence)
	- [ColorSequence](#colorsequence)
	- [NumberRange](#numberrange)
	- [Rect2D](#rect2d)
	- [PhysicalProperties](#physicalproperties)
	- [Color3uint8](#color3uint8)
	- [Int64](#int64)
	- [SharedString](#sharedstring)
- [Data Storage Notes](#data-storage-notes)
	- [Integer Transformations](#integer-transformations)
	- [Interleaved Array](#interleaved-array)
	- [Roblox Float Format](#roblox-float-format)

## Document Conventions
This document assumes a basic understanding of Rust's conventions for numeric types. For example:

- `u16` is an unsigned 16-bit integer
- `i32` is a signed 32-bit integer

Integers are assumed to be little endian and 2's complement unless otherwise specified. Big endian integers and integers with interesting transformations are present in this document and are explicitly noted.

## File Structure
1. File Header
2. Chunks
	1. Zero or one `META` chunks
	2. Zero or one `SSTR` chunks
	3. Zero or more `INST` chunk
	4. Zero or more `PROP` chunks
	5. One `PRNT` chunk
	6. One `END` chunk

## File Header
Every file starts with a 32 byte header.

| Field Name          | Format  | Value                                         |
|:--------------------|:--------|:----------------------------------------------|
| Magic Number        | 8 bytes | Always `<roblox!`                             |
| Signature           | 6 bytes | Always `89 ff 0a 1a 0a`                       |
| Version             | u16     | Always `0`                                    |
| Instance Type Count | i32     | Number of distinct instance types in the file |
| Instance Count      | i32     | Number of instances in the file               |
| Reserved            | 8 bytes | Always `0`                                    |

## Chunks
Every chunk starts with a 16 byte header followed by the chunk's data.

| Field Name          | Format  | Value                                             |
|:--------------------|:--------|:--------------------------------------------------|
| Chunk Name          | 4 bytes | The chunk's name, like `META` or `INST`           |
| Compressed Length   | u32     | Length of the chunk in bytes, if it is compressed |
| Uncompressed Length | u32     | Length of the chunk's data after decompression    |
| Reserved            | 4 bytes | Always `0`                                        |

If **chunk name** is less than four bytes, the remainder is filled with zeros.

If **compressed length** is zero, **chunk data** contains **uncompressed length** bytes of data for the chunk.

If **compressed length** is nonzero, **chunk data** contains an LZ4 compressed block. It is **compressed length** bytes long and will expand to **uncompressed length** bytes when decompressed.

When the **chunk data** is compressed, it is done so using the [LZ4](https://github.com/lz4/lz4) compression algorithm.

When documentation for individual chunks uses the term "chunk data", it refers to **chunk data** after it has been decompressed, if necessary.

### `META` Chunk
The `META` chunk has this layout:

| Field Name                 | Format         | Value                                       |
|:---------------------------|:---------------|:--------------------------------------------|
| Number of Metadata Entries | u32            | The number of metadata entries in the chunk |
| Metadata Entries           | Array(Entries) | The actual metadata entries                 |

Each metadata entry has the following format:

| Field Name     | Format | Value                                    |
|:---------------|:-------|:-----------------------------------------|
| Metadata Key   | String | The metadata key, which should be unique |
| Metadata Value | String | The value for this metadata key          |

The Metadata chunk (`META`) is a map of strings to strings. It represents metadata about the model, such as whether it was authored with `ExplicitAutoJoints` enabled.

There should be zero or one `META` chunks.

Observed metadata entries and their values:

- `ExplicitAutoJoints`: `true` or `false`

### `SSTR` Chunk
The `SSTR` chunk has this layout:

| Field Name          | Format                | Value                                        |
|:--------------------|:----------------------|:---------------------------------------------|
| Version             | u32                   | The version of the `SSTR` chunk (always `0`) |
| Shared String Count | u32                   | The number of SharedStrings in the chunk     |
| Strings             | Array(Shared Strings) | The actual shared string entries             |

A shared string entry looks like this:

| Field Name    | Format   | Value                                                                   |
|:--------------|:---------|:------------------------------------------------------------------------|
| MD5 Hash      | 16 bytes | An [MD5](https://en.wikipedia.org/wiki/MD5) hash of the `Shared String` |
| Shared String | String   | The string that's used by a later `PROP` chunk                          |

The Shared String chunk (`SSTR`) is an array of strings. It's used to reduce the overall size of a file by allowing large strings to be reused in [`PROP`](#prop-chunk) chunks. The `MD5 Hash` isn't used by Roblox Studio when loading the file.

There should be zero or one `SSTR` chunks.

### `INST` Chunk
The `INST` chunk has this layout:

| Field Name         | Format           | Value                                           |
|:-------------------|:-----------------|:------------------------------------------------|
| Type ID            | u32              | An arbitrarily-chosen ID to refer to this type  |
| Type Name          | String           | The type's name, like `Folder` or `Part`        |
| Object Format      | u8               | Always `0` (regular instance) or `1` (service)  |
| Number Instances   | u32              | The number of instances of this type            |
| Instance Referents | Array(Referent)  | The instance referents in the file of this type |
| Service Markers    | Array(u8)        | If **Object Format** is `1`, contains the byte `1` once for each instance in the chunk. Otherwise, not present. |

The Instance chunk (`INST`) defines a type of instance, how many of them there are in this file, and what referent IDs they have.

There should be one `INST` chunk for each type of instance defined.

There are two forms of the `INST` chunk determined by the **Object Format** field:

- `0`: regular
- `1`: service

If the **Object Format** is **regular**, the service markers section will not be present.

If the **Object Format** is **service**, the service markers section contains `1` repeated for the number of instances of that type in the file. If this field is not set, Roblox may create duplicate copies of services, like in [rojo-rbx/rbx-dom#11](https://github.com/rojo-rbx/rbx-dom/issues/11).

**Type ID** must be unique and ideally sorted monotonically among all `INST` chunks. It's used later in the file to refer to this type.

**Type Name** should match the `ClassName` specified on an instance in Roblox.

The length of the **Instance Referents** array must match the **Number of Instances** field.

### `PROP` Chunk
The `PROP` chunk has this layout:

| Field Name    | Format       | Value                                                      |
|:--------------|:-------------|:-----------------------------------------------------------|
| Type ID       | u32          | The type ID assigned in the `INST` chunk                   |
| Property Name | String       | The name of the property, like `CFrame`                    |
| Data Type     | u8           | The [Data Type](#data-types) of this property              |
| Values        | Array(value) | A list of values whose type is determined by **Data Type** |

The property chunk (`PROP`) defines a single property for a single instance type.

There should be one `PROP` chunk per property per instance type.

Because of the shape of this chunk, every instance of a given type must have the same properties specified with the same times. Put another way, if any instance in the file defines a property, all other instances of the same type must also specify that property!

**Type ID** defines the instance that this property applies to as defined in a preceding `INST` chunk.

**Property Name** defines the serializable name of the property. Note that this is not necessarily the same as the name reflected to Lua, which is sometimes referred to as the _canonical name_.

**Data Type** corresponds to a value from [Data Types](#data-types).

**Values** contains an array of values of **Data Type** whose length is the same as the number of instances with the type ID **Type ID**.

### `PRNT` Chunk
The `PRNT` chunk has this layout:

| Field Name        | Format          | Value                                     |
|:------------------|:----------------|:------------------------------------------|
| Version           | u8              | Always `0`                                |
| Number of Objects | u32             | Number of objects described in this chunk |
| Objects           | Array(Referent) | Objects to be parented                    |
| Parents           | Array(Referent) | Parents for objects                       |

The parent chunk (`PRNT`) defines the hierarchy relationship between every instance in the file.

There should be exactly one `PRNT` chunk.

**Version** field should currently always be zero.

**Number of Objects** should be the same as the number of instances in the file header chunk, since each object should have a parent.

**Object Array** and **Parent Array** should both have length equal to **Number of Objects**. The parent of the ID at position *N* in the **Object Array** is a child of the ID at position *N* in the **Parent Array**.

A null parent referent (`-1`) indicates that the object is a root instance. In a place, that means the object is a child of `DataModel`. In a model, that means the object should be placed directly under the object the model is being inserted into.

### `END` Chunk
The `END` chunk has this layout:

| Field Name  | Format  | Value              |
|:------------|:--------|:-------------------|
| Magic Value | 9 bytes | Always `</roblox>` |

The ending chunk (`END`) signifies the end of the file.

The `END` chunk must not be compressed. It is used as a rough form of file validation when uploading places to the Roblox website.

## Data Types

### String
**Type ID `0x01`**

| String |
| ------ |
| String length in bytes (u32) |
| Data |

String data is UTF-8 encoded.

### Bool
**Type ID `0x02`**

The `Bool` type is stored as a single byte. If the byte is `0x00`, the bool is `false`. If it is `0x01`, it is `true`.

When an array of Bools is present, they are stored in sequence.

### Int32
**Type ID `0x03`**

The `Int32` type is stored as a big-endian [transformed 32-bit integer](#integer-transformations).

When an array of Int32s is present, the bytes of the integers are subject to [byte interleaving](#interleaved-array).

### Float32
**Type ID 0x04**

The `Float32` type is stored using the [Roblox float format](#roblox-float-format) and is little-endian. This datatype is also called `float` or `single`.

When an array of Float32s is present, the bytes of the floats are subject to [byte interleaving](#interleaved-array).

### Float64
**Type ID 0x05**

The `Float64` type is stored using the [IEEE-754 format](https://en.wikipedia.org/wiki/Double-precision_floating-point_format) and is little-endian. This datatype is also called `double`.

When an array of Float64s is present, they are in sequence with no transformations.

### UDim
**Type ID 0x06**

### UDim2
**Type ID 0x07**

### Ray
**Type ID 0x08**

### Faces
**Type ID 0x09**

### Axis
**Type ID 0x0A**

### BrickColor
**Type ID 0x0B**

### Color3
**Type ID 0x0C**

### Vector2
**Type ID 0x0D**

### Vector3
**Type ID 0x0E**

### CFrame
**Type ID 0x10**

### Enum
**Type ID 0x12**

### Referent
**Type ID 0x13**

Referents are stored as transformed 32-bit signed integers. A value of `-1` (untransformed) indicates a null referent.

When reading an [Interleaved Array](#interleaved-array) of referents, they should be read accumulatively. In other words, the value of each referent id should be itself, plus its previous value.

Without accumulation, referents read from a file may look like this. This is **incorrect**:

- 1619
- 1
- 4
- 2
- 3
- 5

The **correct** interpretation of this data, with accumulation, is:

- 1619
- 1620
- 1624
- 1626
- 1629
- 1634

### Vector3int16
**Type ID 0x14**

### NumberSequence
**Type ID 0x15**

### ColorSequence
**Type ID 0x16**

### NumberRange
**Type ID 0x17**

### Rect2D
**Type ID 0x18**

### PhysicalProperties
**Type ID 0x19**

### Color3uint8
**Type ID 0x1A**

### Int64
**Type ID 0x1B**

The `Int64` type is stored as a big-endian [transformed 64-bit integer](#integer-transformations).

When an array of Int64s is present, the bytes of the integers are subject to [byte interleaving](#interleaved-array).

### SharedString
**Type ID 0x1C**

SharedStrings are stored as an [Interleaved Array](#interleaved-array) of [Int32s](#int32) that represent indices in the [`SSTR`](#sstr-chunk) string array.

Any property that's a [String](#string) can also be a SharedString.

## Data Storage Notes

### Integer Transformations

Some integers may be subject to a transformation to make them more compressable.

To transform an integer: if `x` greater than or equal to zero, transform it with `2 * x`. Otherwise, use `2 * |x| - 1`. In most compilers this is equivalent to `(x << 1) ^ (x >> 31)` for 32-bit integers. For 64-bit integers, the same format is used but with `63` instead of `31`. 

To untransform one: if `x` is divisble by 2, untransform it with `x / 2`. Otherwise, use `-(x + 1) / 2`. This is equivalent to `(x >> 1) ^ -(x & 1)`.

Untransforming with bitwise operators requires casting to an unsigned integer in some cases because `x >> 1` will result in a negative number if `x` is negative.

### Interleaved Array
Arrays of many types in property data have their bytes interleaved.

For example, an array of 4 bit integers normally represented as:

|||||||||||||
|--|--|--|--|--|--|--|--|--|--|--|--|
|**A0**|**A1**|**A2**|**A3**|B0|B1|B2|B3|C0|C1|C2|C3|

Would become, after interleaving:

|||||||||||||
|--|--|--|--|--|--|--|--|--|--|--|--|
|**A0**|B0|C0|**A1**|B1|C1|**A2**|B2|C2|**A3**|B3|C3|

Note that arrays of integers are generally subject to both interleaving and integer transformation.

### Roblox Float Format

Some data types do not follow the [IEEE-754 standard](https://en.wikipedia.org/wiki/Single-precision_floating-point_format) format for 32-bit floating point numbers. Instead, they use a proprietary format where the sign bit is after the mantissa.

| Format   | Bit Layout                            |
|:---------|:--------------------------------------|
| Standard | `seeeeeee emmmmmmm mmmmmmmm mmmmmmmm` |
| Roblox   | `eeeeeeee mmmmmmmm mmmmmmmm mmmmmmms` |

Where `s` is the sign bit, `e` is an exponent bit, and `m` is a mantissa bit.

As a practical example, below is a comparison of how `-0.15625` is stored:

| Format   | Binary View                           | Byte View     |
|:---------|:--------------------------------------|:--------------|
| Standard | `10111110 00100000 00000000 00000000` | `be 20 00 00` |
| Roblox   | `01111100 01000000 00000000 00000001` | `7c 40 00 01` |