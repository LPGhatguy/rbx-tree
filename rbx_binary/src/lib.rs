//! Implementation of Roblox's binary model (rbxm) and place (rbxl) file
//! formats.

#![deny(missing_docs)]

mod cframe;
mod chunk;
mod core;
mod deserializer;
mod serializer;
mod types;

#[cfg(any(test, feature = "unstable_text_format"))]
mod text_deserializer;

#[cfg(test)]
mod tests;

use std::io::{Read, Write};

use rbx_dom_weak::{types::Ref, WeakDom};

use crate::{deserializer::decode, serializer::encode};

/// An unstable textual format that can be used to debug binary models.
#[cfg(feature = "unstable_text_format")]
pub mod text_format {
    pub use crate::text_deserializer::*;
}

pub use crate::{deserializer::Error as DecodeError, serializer::Error as EncodeError};

/// Decodes an binary format model or place from something that implements the
/// `std::io::Read` trait.
pub fn from_reader_default<R: Read>(reader: R) -> Result<WeakDom, DecodeError> {
    decode(reader)
}

/// Serializes a subset of the given DOM to a binary format model or place,
/// writing to something that implements the `std::io::Write` trait.
pub fn to_writer_default<W: Write>(
    writer: W,
    dom: &WeakDom,
    refs: &[Ref],
) -> Result<(), EncodeError> {
    encode(dom, refs, writer)
}
