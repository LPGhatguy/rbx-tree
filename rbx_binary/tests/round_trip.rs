use std::{fs, path::Path};

use rbx_dom_test::{InstanceBuilder, TreeViewer};

use rbx_binary::{decode, encode};

#[test]
fn just_folder() {
    test_round_trip("from-studio/default-inserted-folder.rbxm");
}

#[test]
fn folder_with_script() {
    test_round_trip("folder-with-script.rbxm");
}

#[test]
fn hierarchy_with_refs() {
    test_round_trip("hierarchy-with-refs.rbxm");
}

/// Helper method to test both reading and writing using only rbx_binary.
fn test_round_trip(test_path: &str) {
    let _ = env_logger::try_init();

    let input_path = Path::new("test-files").join(test_path);
    let test_name = input_path.file_name().unwrap().to_str().unwrap();

    let initial_contents = fs::read(&input_path).unwrap();

    let mut initial_tree = InstanceBuilder::new("Folder").build();

    let root_id = initial_tree.get_root_id();
    decode(&mut initial_tree, root_id, initial_contents.as_slice()).unwrap();

    let initial_viewed = TreeViewer::new().view_children(&initial_tree);
    insta::assert_yaml_snapshot!(test_name, initial_viewed);

    let root_instance = initial_tree.get_instance(root_id).unwrap();
    let root_children = root_instance.get_children_ids();
    let mut roundtrip_contents = Vec::new();
    encode(&initial_tree, root_children, &mut roundtrip_contents).unwrap();

    let mut roundtrip_tree = InstanceBuilder::new("Folder").build();

    let root_id = roundtrip_tree.get_root_id();
    decode(&mut roundtrip_tree, root_id, roundtrip_contents.as_slice()).unwrap();

    let roundtrip_viewed = TreeViewer::new().view_children(&roundtrip_tree);
    insta::assert_yaml_snapshot!(test_name, roundtrip_viewed);
}
