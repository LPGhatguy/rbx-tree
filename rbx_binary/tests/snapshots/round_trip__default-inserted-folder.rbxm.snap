---
source: rbx_binary/tests/round_trip.rs
expression: initial_viewed
---
- id: id-0
  name: Folder
  class_name: Folder
  properties:
    AttributesSerialize:
      Type: BinaryString
      Value: ""
    Tags:
      Type: BinaryString
      Value: ""
  children: []
